#!/usr/bin/env python3
PACKAGE = "manicapture"

from dynamic_reconfigure.parameter_generator_catkin import *
#initial ros and generators
gen = ParameterGenerator()
#gen.add(name , type, level, description, default, min, max)

function_node_enum_adc = gen.enum([
	gen.const("Mono", int_t, 8, "8"), gen.const("Light", int_t, 9, "9"),
	gen.const("Low", int_t, 10, "10"), gen.const("Medium", int_t, 11, "11"),
					   gen.const("High", int_t, 12, "12")], "switch adc_resolution")
function_node_enum_sd = gen.enum([ 
	gen.const("50", int_t, 50, "50"),
	gen.const("100", int_t, 100, "100"),
	gen.const("150", int_t, 150, "150"),
	gen.const("200",  int_t, 200, "200"),
	gen.const("300",  int_t, 300, "300"),
	gen.const("400",  int_t, 400, "400"),
	gen.const("500",  int_t, 500, "500"),
	gen.const("1000",  int_t, 1000, "1000"),
	gen.const("2000",  int_t, 2000, "2000"),
	gen.const("10000",  int_t, 10000, "10000")],
	"switch sense_delay")
function_node_enum_id = gen.enum([gen.const("close", int_t, 0, "close"),
					   gen.const("open",  int_t, 1, "open")], "switch indicator led")
function_node_enum_gt = gen.enum([gen.const("Mono_8x8_1", int_t, 0, "Mono"), 
								gen.const("Dense_8x8_5", int_t, 1, "DenseMedium"),
					   			gen.const("DenseFull_8x8_9", int_t, 2, "DenseFull"),
								gen.const("Normal_Tight_6x6_5", int_t, 3, "NormalMedium"),
								gen.const("Normal_Tight_6x6_9", int_t, 4, "NormalFull")], 
                                "set the sensing grid type")
function_node_enum_sr = gen.enum([ 
	gen.const("None", int_t, -1, "None"),
	gen.const("TACT0", int_t, 0, "TACT0"),
	gen.const("TACT1", int_t, 1, "TACT1"),
	gen.const("TACT2",  int_t, 2, "TACT2"),
	gen.const("TACT3",  int_t, 3, "TACT3")],
	"show chosen tactile photo image")
function_node_enum_hv = gen.enum([ 
	gen.const("VERSION_1", int_t, 1, "1"),
	gen.const("VERSION_2",  int_t, 2, "2")],
	"hardware version")

gen.add("TACT0", bool_t, 2, "tactile hardware 0",  True)
gen.add("TACT1", bool_t, 2, "tactile hardware 1",  False)
gen.add("TACT2", bool_t, 2, "tactile hardware 2",  False)
gen.add("TACT3", bool_t, 2, "tactile hardware 3",  False)
gen.add("adc_resolution", int_t, 1, "set analog read resolution of photo (8-12)",  
		12, 8, 12, edit_method=function_node_enum_adc)
gen.add("sense_delay", int_t, 1, "set waiting delay us of photo reading (50-10000)",  
		200, 50, 10000, edit_method=function_node_enum_sd)
gen.add("grid_type", int_t, 2, "set the sensing grid type", 2, 0, 4, edit_method=function_node_enum_gt)
gen.add("update_range", int_t, 1, "set the sensing sensitivity", 2, 0, 32)
gen.add("sequence", int_t, 1, "set the sensing sequence", 0, 0, 1)
gen.add("hardware", int_t, 1, "set the hardware version", 2, 1, 2, edit_method=function_node_enum_hv)
gen.add("indicator", int_t, 1, "set indicator state", 1, 0, 1, edit_method=function_node_enum_id)
gen.add("showrosimage", int_t, 3, "show chosen tactile photo image", 0, -1, 3, edit_method=function_node_enum_sr)

#exit and file creation
exit(gen.generate(PACKAGE, "manicapture", "machineparams"))

